# Project Description
Diamonds as we all know are very expensive. But the range of cost can vary significantly for similarly sized diamonds. Using variables describing diamond quality I am going to build a model to predict the price of the diamond.

# Data Preprocessing
## Importing the Libraries

In [203]:
import numpy as np
import pandas as pd

## Importing Dataset

In [204]:
dataset = pd.read_csv('diamonds.csv')

## Showing the Dataset in a Table

In [205]:
pd.DataFrame(dataset)

Unnamed: 0  carat        cut color clarity  depth  table  price     x  \
0               1   0.23      Ideal     E     SI2   61.5   55.0    326  3.95   
1               2   0.21    Premium     E     SI1   59.8   61.0    326  3.89   
2               3   0.23       Good     E     VS1   56.9   65.0    327  4.05   
3               4   0.29    Premium     I     VS2   62.4   58.0    334  4.20   
4               5   0.31       Good     J     SI2   63.3   58.0    335  4.34   
...           ...    ...        ...   ...     ...    ...    ...    ...   ...   
53935       53936   0.72      Ideal     D     SI1   60.8   57.0   2757  5.75   
53936       53937   0.72       Good     D     SI1   63.1   55.0   2757  5.69   
53937       53938   0.70  Very Good     D     SI1   62.8   60.0   2757  5.66   
53938       53939   0.86    Premium     H     SI2   61.0   58.0   2757  6.15   
53939       53940   0.75      Ideal     D     SI2   62.2   55.0   2757  5.83   

          y     z  
0      3.98  2.43  
1      3.84  2.31  
2      4.07  2.31  
3      4.23  2.63  
4      4.35  2.75  
...     ...   ...  
53935  5.76  3.50  
53936  5.75  3.61  
53937  5.68  3.56  
53938  6.12  3.74  
53939  5.87  3.64  

[53940 rows x 11 columns]

## Drop Unwanted Columns
The column that will be dropped is the first one which has no name "Unnamed".  

In [206]:
dataset = dataset.drop(['Unnamed: 0'], axis=1)

## Seperate The Input and Output
Here, we put the independent variables in X and the dependent variable in Y. 

In [207]:
X = dataset.drop(['price'], axis=1)
Y = dataset['price']

## Encoding Categorical Data
The categorical data that I will be one hot encoding are columns "cut", "color", and "clarity".

In [208]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder

# instantiate OneHotEncoder object
ohe = OneHotEncoder()
# instantiate ColumnTransformer object
ct = ColumnTransformer(transformers=[('encoder1', ohe, ['cut', 'color', 'clarity'])], remainder='passthrough')
# transform data in array X
X = ct.fit_transform(X)

array([0.23, 0.21, 0.23, ..., 0.7 , 0.86, 0.75])

## Feature Scaling
The independent variables that I will be feature scaling are all "carar", "depth", "table", "price", "x", "y", and "z".

In [209]:
from sklearn.preprocessing import StandardScaler

# instantiate StandardScalar object
sc = StandardScaler()
# scale data
X[:, 20:] = sc.fit_transform(X[:, 20:])

## Splitting the Dataset into the Training set and Test set

In [210]:
from sklearn.model_selection import train_test_split

# split dataset, train:test split
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size = 0.2, random_state = 1)

# Regression Model Analysis
## Linear Regression
Train the regression model.

In [211]:
from sklearn import linear_model

# instantiate regression object
linear = linear_model.LinearRegression()
# train model
linear = linear.fit(X_train, y_train)

Print out the models accuracy.

In [212]:
linear.score(X_test, y_test)

0.9217786442329299

# Conclusion
The models accuracy using linear regression is approximately 92%!